service/frontend/http_api_server.go

```diff
@@ -0,0 +1,438 @@
+	h.server.ConnContext = func(ctx context.Context, c net.Conn) context.Context {
+		return context.WithValue(ctx, httpRemoteAddrContextKey{}, c)
+	}
+
+	// We want to set ReadTimeout and WriteTimeout as max idle (and IdleTimeout
+	// defaults to ReadTimeout) to ensure that a connection cannot hang over that
+	// amount of time.
+	h.server.ReadTimeout = serviceConfig.KeepAliveMaxConnectionIdle()
+	h.server.WriteTimeout = serviceConfig.KeepAliveMaxConnectionIdle()
+
+	success = true
+	return h, nil
+}
+
+// Serve serves the HTTP API and does not return until there is a serve error or
+// GracefulStop completes. Upon graceful stop, this will return nil. If an error
+// is returned, the message is clear that it came from the HTTP API server.
+func (h *HTTPAPIServer) Serve() error {
+	err := h.server.Serve(h.listener)
+	// If the error is for close, we have to wait for the shutdown to complete and
+	// we don't consider it an error
+	if err == http.ErrServerClosed {
+		<-h.stopped
+		err = nil
+	}
+	// Wrap the error to be clearer it's from the HTTP API
+	if err != nil {
+		return fmt.Errorf("HTTP API serve failed: %w", err)
+	}
+	return nil
+}
+
+// GracefulStop stops the HTTP server. This will first attempt a graceful stop
+// with a drain time, then will hard-stop. This will not return until stopped.
+func (h *HTTPAPIServer) GracefulStop() {
+	// We try a graceful stop for the amount of time we can drain, then we do a
+	// hard stop
+	shutdownCtx, _ := context.WithTimeout(context.Background(), h.shutdownDrainTime)
+	// We intentionally ignore this error, we're gonna stop at this point no
+	// matter what. This closes the listener too.
+	_ = h.server.Shutdown(shutdownCtx)
+	_ = h.server.Close()
+	close(h.stopped)
+}
+
+func (h *HTTPAPIServer) serveHTTP(w http.ResponseWriter, r *http.Request) {
+	// Limit the request body to max gRPC size. This is hardcoded to 4MB at the
+	// moment using gRPC's default at
+	// https://github.com/grpc/grpc-go/blob/0673105ebcb956e8bf50b96e28209ab7845a65ad/server.go#L58.
+	// Max header bytes is defaulted by Go to 1MB.
+	r.Body = http.MaxBytesReader(w, r.Body, 1024*1024*4)
+
+	h.logger.Debug(
+		"HTTP API call",
+		tag.NewStringTag("http-method", r.Method),
+		tag.NewAnyTag("http-url", r.URL),
+	)
+
+	// Need to change the accept header based on whether pretty and/or
+	// noPayloadShorthand are present
+	var acceptHeaderSuffix string
+	if _, ok := r.URL.Query()["pretty"]; ok {
+		acceptHeaderSuffix += "+pretty"
+	}
+	if _, ok := r.URL.Query()["noPayloadShorthand"]; ok {
+		acceptHeaderSuffix += "+no-payload-shorthand"
+	}
+	if acceptHeaderSuffix != "" {
+		r.Header.Set("Accept", "application/json"+acceptHeaderSuffix)
+	}
+
+	// Put the TLS info on the peer context
+	if r.TLS != nil {
+		var addr net.Addr
+		if conn, _ := r.Context().Value(httpRemoteAddrContextKey{}).(net.Conn); conn != nil {
+			addr = conn.RemoteAddr()
+		}
+		r = r.WithContext(peer.NewContext(r.Context(), &peer.Peer{
```

Comment:
nice. this means we don't have to change the authorization interceptor at all?

Reply:
Exactly. And I have tests proving it.